### ISSUE TITLE
C++20 spans and range algorithms give erroneous Intellisense errors

### ISSUE THREAD

**Issue by 0xjc** [microsoft/vscode-cpptools#10024]
### Environment

- OS and Version: Ubuntu 20.04.5 LTS
- VS Code Version: 1.73.0-insider
- C/C++ Extension Version: v1.13.2

When using gcc-10 in C++20 mode, Intellisense will often give errors on conversions of contiguous ranges to `std::span`, and also when using most algorithms under `std::ranges`. This code compiles and works properly under `g++-10 -std=c++20`.

### Bug Summary and Steps to Reproduce

See code

### Code sample and Logs

Code sample:

```
#include <algorithm>
#include <iostream>
#include <ranges>
#include <span>
#include <vector>

void f_0(std::span<const int> x) { std::cout << x.size() << "\n"; }

void f(const std::vector<int>& x) { f_0(x); }
//                                      ^
// no suitable user-defined conversion from "const std::vector<int,
// std::allocator<int>>" to "std::span<const int, 18446744073709551615UL>"
// exists C/C++(312)

void g(const std::vector<int>& x) { std::cout << std::ranges::max(x) << "\n"; }
//                                               ^^^
// no instance of overloaded function "std::ranges::__max_fn::operator()"
// matches the argument list C/C++(304)
// argument types are: (const std::vector<int, std::allocator<int>>)
// object type is: const std::ranges::__max_fn

int main() {
  auto x = std::vector{3, 4, 5};
  f(x);
  g(x);
}
```

`c_cpp_properties.json`:

```
{
  "configurations": [
      {
          "name": "Linux",
          "intelliSenseMode": "gcc-x64",
          "cppStandard": "c++20",
          "compilerPath": "/usr/bin/gcc-10"
      }
  ],
  "version": 4
}
```

Diagnostics:
```
-------- Diagnostics - 10/17/2022, 1:35:18 PM
Version: 1.13.2
Current Configuration:
{
    "name": "Linux",
    "intelliSenseMode": "gcc-x64",
    "cppStandard": "c++20",
    "compilerPath": "/usr/bin/gcc-10",
    "compilerPathIsExplicit": true,
    "cStandardIsExplicit": false,
    "cppStandardIsExplicit": true,
    "intelliSenseModeIsExplicit": true,
    "mergeConfigurations": false,
    "browse": {
        "limitSymbolsToIncludedHeaders": true
    }
}
...
Translation Unit Configurations:
[ /home/jcai/test/test.cpp ]:
    Process ID: 1264701
    Memory Usage: 770 MB
    Compiler Path: /usr/bin/gcc-10
    Includes:
        /usr/include/c++/10
        /usr/include/x86_64-linux-gnu/c++/10
        /usr/include/c++/10/backward
        /usr/lib/gcc/x86_64-linux-gnu/10/include
        /usr/local/include
        /usr/include/x86_64-linux-gnu
        /usr/include
    Standard Version: c++20
    IntelliSense Mode: linux-gcc-x64
    Other Flags:
        --g++
        --gnu_version=100300
```

**Comment by hwmaier** [comment]
Experiencing similar issue. Despite configuring `"cppStandard": "c++20"` in `c_cpp_properties.json`, the parser flags `namespace "std" has no member "span"` errors.  Code compiles w/o warning with MSVC.

**Comment by Colengms** [comment]
Hi @0xjc .  I'm able to reproduce the issue.  I've opened an internal issue (1656876) against the IntelliSense library.  We share the IntelliSense implementation with VS.

**Comment by Grant12311** [comment]
Hi! I'm having a similar issue. I get this error when using std::ranges::any_of and similar messages with other ranges stuff. The code compiles just fine with GCC 11 though.
```
no instance of overloaded function "std::ranges::__all_of_fn::operator()" matches the argument list
```

**Comment by Grant12311** [comment]
Yeah last I checked this issue is still ongoing so an update would be appreciated.

**Comment by mwhds97** [comment]
I have the same issue, with GCC 12

**Comment by sean-mcmanus** [comment]
No progress has been made yet, but I've pinged the team and mentioned the 9 or so upvotes or user comments.

**Comment by davidBronicki** [comment]
Also having this issue. Probably related: intellisense believes `ranges::range<vector<double>>` evaluates to false. I've tried this with clang 15 and 17.

**Comment by Jesseeee** [comment]
Having the same issue with gcc 12.2 trying to use any std::ranges functions, gives such errors
"message": "no instance of overloaded function \"std::ranges::__find_fn::operator()\" matches the argument list",
"message": "argument types are: (std::vector<SessionId, std::allocator<SessionId>>, const SessionId)",
"message": "object type is: const std::ranges::__find_fn",

**Comment by marxin** [comment]
I can confirm it works for me with the update extension and my test-case (`mold` linker)! Thank you for the fix.

**Comment by janderudder** [comment]
Is this fixed in vs code ? I use it on Windows and it auto updates regularly but Intellisense still reports an error when converting from std::vector using std::span constructor.

**Comment by sean-mcmanus** [comment]
@janderudder It's fixed with 1.15.1 (pre-release), but there are other span/range-related bugs that are not fixed yet which we have other issues tracking, so you could be hitting those or you could file a new issue with more repro details.

**Comment by janderudder** [comment]
Thank you @sean-mcmanus. I think I'm in the most basic situation of getting, for instance, `no suitable user-defined conversion from "const std::vector<std::string, std::allocator<std::string>>" to "std::span<const std::string, 18446744073709551615UL>" exists [C/C++(312)]` when simply using a `std::vector` where a `std::span` is expected, or constructing a span by assigning a vector (which of course compiles without issue).

I just noticed that the error goes away when disabling Microsoft "C/C++ for Visual Studio Code" extension. I have vscode 1.77.0.

**Comment by sean-mcmanus** [comment]
@janderudder Can you file a new bug report on that with a code sample and info like the compiler/OS used? I'm not sure if that particular case tracked by an existing issue or not.

**Comment by sean-mcmanus** [comment]
Fixed with https://github.com/microsoft/vscode-cpptools/releases/tag/v1.15.4, and more range fixes are going to be in 1.16.0 (https://github.com/microsoft/vscode-cpptools/issues/8039).
